# Performing Data Wrangling on San Francisco Crime Data

San Francisco's Government keeps a record of police reports filed in their Crime Data Warehouse hosted at https://datasf.org/. Having crimes available online rather than passing around files of papers can make searching for past crimes and analyzing crime patterns from the data set much simpler.

In the years of 2008-2013, "the district attorney’s domestic violence team dropped about 72 percent of all domestic violence criminal cases before they reached court" (https://sfpublicpress.org/news/2013-06/domestic-violence-case-that-spurred-san-francisco-reforms-comes-to-a-close). Because of the poor record-keeping of the San Francisco Police Department, domestic crimes were falling through without any action being taken. This is where the Crime Data Warehouse plays such an important role in recording and maintaining an organized data set of crime reports to be considered. 

I plan on using two data sets: one containing reports from January 2003-May 2018 and the other containing reports after the Crime Data Warehouse update by the San Francisco Police Department at the start of 2018 (which is updated daily).

In [2]:
# Import necessary modules
import pandas as pd

## *TABLE 1: Police Reports from January 2003 - May 2018*

In [3]:
file_1 = "2003-2018.csv"
table_1 = pd.read_csv(file_1)
table_1.shape

(2215024, 33)

In [4]:
table_1.head()

IncidntNum        Category                      Descript DayOfWeek  \
0   180362289   VEHICLE THEFT             STOLEN MOTORCYCLE   Tuesday   
1   180360948    NON-CRIMINAL  AIDED CASE, MENTAL DISTURBED   Tuesday   
2   180360879  OTHER OFFENSES              PAROLE VIOLATION   Tuesday   
3   180360879  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   Tuesday   
4   180360879  OTHER OFFENSES             TRAFFIC VIOLATION   Tuesday   

         Date   Time PdDistrict      Resolution  \
0  05/15/2018  10:30   SOUTHERN            NONE   
1  05/15/2018  04:14   SOUTHERN            NONE   
2  05/15/2018  02:01    MISSION  ARREST, BOOKED   
3  05/15/2018  02:01    MISSION  ARREST, BOOKED   
4  05/15/2018  02:01    MISSION  ARREST, BOOKED   

                         Address           X  ...  \
0         700 Block of TEHAMA ST -122.411912  ...   
1  MARKET ST / SOUTH VAN NESS AV -122.419258  ...   
2              CAPP ST / 21ST ST -122.417813  ...   
3              CAPP ST / 21ST ST -122.417813  ...   
4              CAPP ST / 21ST ST -122.417813  ...   

   :@computed_region_n4xg_c4py :@computed_region_4isq_27mq  \
0                          1.0                         NaN   
1                          1.0                         NaN   
2                          NaN                        15.0   
3                          NaN                        15.0   
4                          NaN                        15.0   

   :@computed_region_fcz8_est8  :@computed_region_pigm_ib2e  \
0                          1.0                          NaN   
1                          1.0                          NaN   
2                          3.0                         15.0   
3                          3.0                         15.0   
4                          3.0                         15.0   

   :@computed_region_9jxd_iqea  :@computed_region_6pnf_4xz7  \
0                          NaN                          2.0   
1                          8.0                          2.0   
2                          NaN                          2.0   
3                          NaN                          2.0   
4                          NaN                          2.0   

   :@computed_region_6ezc_tdp2  :@computed_region_h4ep_8xdi  \
0                          NaN                          NaN   
1                          1.0                          1.0   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   :@computed_region_nqbw_i6c3  :@computed_region_2dwj_jsy4  
0                          1.0                          NaN  
1                          1.0                          NaN  
2                          3.0                          NaN  
3                          3.0                          NaN  
4                          3.0                          NaN  

[5 rows x 33 columns]

## *TABLE 2: Police Reports from January 2018 - Present*

In [5]:
file_2 = "2018-present.csv"
table_2 = pd.read_csv(file_2)
table_2.shape

(246942, 34)

In [6]:
table_2.head()

Incident Datetime Incident Date Incident Time  Incident Year  \
0  2018/07/18 01:30:00 PM    2018/07/18         13:30           2018   
1  2019/08/05 03:29:00 AM    2019/08/05         03:29           2019   
2  2019/04/08 05:25:00 AM    2019/04/08         05:25           2019   
3  2019/06/05 02:00:00 PM    2019/06/05         14:00           2019   
4  2019/08/05 05:22:00 AM    2019/08/05         05:22           2019   

  Incident Day of Week         Report Datetime       Row ID  Incident ID  \
0            Wednesday  2018/07/18 01:31:00 PM  69250964070       692509   
1               Monday  2019/08/05 03:29:00 AM  83072007043       830720   
2               Monday  2019/04/13 01:34:00 PM  79165671000       791656   
3            Wednesday  2019/06/05 02:30:00 PM  81006072000       810060   
4               Monday  2019/08/05 05:22:00 AM  83083709330       830837   

   Incident Number   CAD Number  ...   Longitude  \
0        180536729  182001522.0  ...         NaN   
1        190561302          NaN  ...         NaN   
2        196076240          NaN  ...         NaN   
3        190416337  191610728.0  ... -122.422464   
4        190385817          NaN  ...         NaN   

                                      point SF Find Neighborhoods  \
0                                       NaN                   NaN   
1                                       NaN                   NaN   
2                                       NaN                   NaN   
3  (37.78268536745206, -122.42246374465972)                 100.0   
4                                       NaN                   NaN   

   Current Police Districts Current Supervisor Districts  \
0                       NaN                          NaN   
1                       NaN                          NaN   
2                       NaN                          NaN   
3                       4.0                         11.0   
4                       NaN                          NaN   

  Analysis Neighborhoods HSOC Zones as of 2018-06-05 OWED Public Spaces  \
0                    NaN                         NaN                NaN   
1                    NaN                         NaN                NaN   
2                    NaN                         NaN                NaN   
3                   39.0                         NaN                NaN   
4                    NaN                         NaN                NaN   

  Central Market/Tenderloin Boundary Polygon - Updated  \
0                                                NaN     
1                                                NaN     
2                                                NaN     
3                                                NaN     
4                                                NaN     

   Parks Alliance CPSI (27+TL sites)  
0                                NaN  
1                                NaN  
2                                NaN  
3                                NaN  
4                                NaN  

[5 rows x 34 columns]

## Consolidating the Data

Firstly, I want to clearly define the two different data sets that I am using. **Table 1** contains police incident reports that date back to 2003, and continue until May 15, 2018. The reason that this data set was discontinued was due to CABLE, the legacy mainframe that was used to maintain the data, being extremeley prone to issues relating to delays and data accessibility. **Table 2** is the updated version with a detailed API describing the police incident report information. I plan on using **Table 1** for dates from the beginning of 2003 until the end of 2017 and **Table 2** from the start of 2018 to present day. I will be fitting the old table to the new table and appending the rows to create a single table consisting of all police incidents.

The columns in **Table 1** contain similar information to **Table 2.** Some columns in **Table 1**, such as *"IncidntNum"*, have corresponding columns in **Table 2** (*"Incident Number"*). I want to update the **Table 1** column names to fit the data in **Table 2** and append the incident reports of **Table 1** to **Table 2** in order to combine their data into one dataframe. This will make accessing and manipulating report data easier since all reports will be held in a single dataframe.

Steps I am taking:
    1. Drop irrelevant columns in **Table 1**
    2. Reverse rows in **Table 1** to get data sorted by increasing date
    3. Rename columns in **Table 1** to their corresponding column names in **Table 2** to prepare for merging
    4. Sorting rows in **Table 2** by *Incident Datetime*
    5. Perform an outer merge on the two tables to get a single dataframe with all incident reports sorted by date

In [7]:
# Drop the 'computed region' columns in the initial table_1 dataframe, as they are unecessary and irrelevant
table_1.drop(table_1.iloc[:, 17:], inplace=True, axis=1)

In [8]:
table_1.columns

Index(['IncidntNum', 'Category', 'Descript', 'DayOfWeek', 'Date', 'Time',
       'PdDistrict', 'Resolution', 'Address', 'X', 'Y', 'Location', 'PdId',
       'SF Find Neighborhoods', 'Current Police Districts',
       'Current Supervisor Districts', 'Analysis Neighborhoods'],
      dtype='object')

Next, I am reversing the order of the rows of **Table 1** so that the resulting table is sorted by increasing date. To maintain consistency, I reset the index to begin at 0 and filter out all rows who's date are in 2018 so that **Table 1** ends at the end of 2017.

In [9]:
# Reversing the order of table_1 and resetting the index to start at 0
new_table_1 = table_1.iloc[::-1].reset_index(drop=True)

# Filtering table_1 to not have any values from 2018
new_table_1 = new_table_1[~new_table_1.Date.str.contains("2018")]
new_table_1.head()

IncidntNum                Category                               Descript  \
0   100087859  SEX OFFENSES, FORCIBLE               ANNOY OR MOLEST CHILDREN   
1    91139945  SEX OFFENSES, FORCIBLE                         SEXUAL BATTERY   
2    81360718            EMBEZZLEMENT  EMBEZZLEMENT, GRAND THEFT BY EMPLOYEE   
3    81313927          SUSPICIOUS OCC                  SUSPICIOUS OCCURRENCE   
4    81183231  FORGERY/COUNTERFEITING    FORGERY, DRIVERS LICENSE OR ID-CARD   

   DayOfWeek        Date   Time PdDistrict  \
0  Wednesday  01/01/2003  00:01   SOUTHERN   
1  Wednesday  01/01/2003  00:01    BAYVIEW   
2  Wednesday  01/01/2003  00:01   RICHMOND   
3  Wednesday  01/01/2003  00:01    MISSION   
4  Wednesday  01/01/2003  00:01   NORTHERN   

                               Resolution                     Address  \
0        COMPLAINANT REFUSES TO PROSECUTE      800 Block of BRYANT ST   
1  DISTRICT ATTORNEY REFUSES TO PROSECUTE  200 Block of BRIDGEVIEW DR   
2                                    NONE    2900 Block of PACIFIC AV   
3                                    NONE    500 Block of GUERRERO ST   
4                                    NONE   3000 Block of FILLMORE ST   

            X          Y                               Location  \
0 -122.403405  37.775421   (37.775420706711, -122.403404791479)   
1 -122.397590  37.734209  (37.7342093398821, -122.397590096788)   
2 -122.443835  37.792137  (37.7921365604231, -122.443835267233)   
3 -122.424037  37.762751  (37.7627513875316, -122.424036887526)   
4 -122.435517  37.797573  (37.7975726990109, -122.435517480769)   

             PdId  SF Find Neighborhoods  Current Police Districts  \
0  10008785914010                   32.0                       1.0   
1   9113994504144                   87.0                       2.0   
2   8136071810045                  102.0                       8.0   
3   8131392764070                   37.0                       3.0   
4   8118323109120                   15.0                       4.0   

   Current Supervisor Districts  Analysis Neighborhoods  
0                          10.0                    34.0  
1                           9.0                     1.0  
2                           6.0                    30.0  
3                           5.0                    20.0  
4                           6.0                    13.0

Below, I am renaming the columns in **Table 1** to their corresponding column names in **Table 2** to make merging the two dataframes easier in the future. I am also storing the resulting dataframe to **new_table_1** as reference. 

In [10]:
# Rename the columns in table_1 to have the same names as their corresponding columns in table_2
new_table_1 = new_table_1.rename(columns={'IncidntNum': 'Incident Number',
                      'Category': 'Incident Category',
                      'Descript': 'Incident Description',
                      'DayOfWeek': 'Incident Day of Week',
                      'Date': 'Incident Date',
                      'Time': 'Incident Time',
                      'PdDistrict': 'Police District',
                      'X': 'Longitude',
                      'Y': 'Latitude',
                      'Location': 'point',
                      'PdId': 'Row ID'})
new_table_1.tail()

Incident Number Incident Category  \
2168351        186005077     LARCENY/THEFT   
2168352        180000069       WEAPON LAWS   
2168353        180000069          BURGLARY   
2168354        180000069           ASSAULT   
2168355        180000417           ASSAULT   

                                   Incident Description Incident Day of Week  \
2168351                         PETTY THEFT OF PROPERTY               Sunday   
2168352  EXHIBITING DEADLY WEAPON IN A THREATING MANNER               Sunday   
2168353     BURGLARY OF APARTMENT HOUSE, FORCIBLE ENTRY               Sunday   
2168354                 AGGRAVATED ASSAULT WITH A KNIFE               Sunday   
2168355                                         BATTERY               Sunday   

        Incident Date Incident Time Police District      Resolution  \
2168351    12/31/2017         23:55         MISSION            NONE   
2168352    12/31/2017         23:56      TENDERLOIN  ARREST, BOOKED   
2168353    12/31/2017         23:56      TENDERLOIN  ARREST, BOOKED   
2168354    12/31/2017         23:56      TENDERLOIN  ARREST, BOOKED   
2168355    12/31/2017         23:58        SOUTHERN            NONE   

                          Address   Longitude   Latitude  \
2168351  500 Block of VALENCIA ST -122.421876  37.764089   
2168352     500 Block of JONES ST -122.412999  37.786277   
2168353     500 Block of JONES ST -122.412999  37.786277   
2168354     500 Block of JONES ST -122.412999  37.786277   
2168355    800 Block of BRYANT ST -122.403405  37.775421   

                                            point          Row ID  \
2168351  (37.76408889445322, -122.42187648849193)  18600507706372   
2168352  (37.78627745916602, -122.41299907500884)  18000006912030   
2168353  (37.78627745916602, -122.41299907500884)  18000006905011   
2168354  (37.78627745916602, -122.41299907500884)  18000006904012   
2168355    (37.775420706711, -122.40340479147905)  18000041704134   

         SF Find Neighborhoods  Current Police Districts  \
2168351                   37.0                       3.0   
2168352                   20.0                       5.0   
2168353                   20.0                       5.0   
2168354                   20.0                       5.0   
2168355                   32.0                       1.0   

         Current Supervisor Districts  Analysis Neighborhoods  
2168351                           5.0                    20.0  
2168352                          10.0                    36.0  
2168353                          10.0                    36.0  
2168354                          10.0                    36.0  
2168355                          10.0                    34.0

As the last step before merging the two dataframes, I sort **Table 2** on the column *"Incident Datetime"* to maintain a similar format as **new_table_1.** The resulting table, **new_table_2** is also sorted by the date that the crime occured and has the index values reset to begin at 0. These crime reports will begin at the start of 2018 and will continue to the most recent update of the database.

In [11]:
# Sort table_2 on "Incident Datetime" and reset index starting at 0
new_table_2 = table_2.sort_values("Incident Datetime").reset_index(drop=True)
new_table_2.head()

Incident Datetime Incident Date Incident Time  Incident Year  \
0  2018/01/01 01:00:00 AM    2018/01/01         01:00           2018   
1  2018/01/01 01:00:00 AM    2018/01/01         01:00           2018   
2  2018/01/01 01:00:00 AM    2018/01/01         01:00           2018   
3  2018/01/01 01:00:00 AM    2018/01/01         01:00           2018   
4  2018/01/01 01:00:00 AM    2018/01/01         01:00           2018   

  Incident Day of Week         Report Datetime       Row ID  Incident ID  \
0               Monday  2018/01/03 12:58:00 PM  61977304033       619773   
1               Monday  2018/01/02 11:08:00 AM  61958471000       619584   
2               Monday  2018/02/09 11:25:00 AM  63428271000       634282   
3               Monday  2018/01/11 02:39:00 PM  62367106372       623671   
4               Monday  2018/02/26 09:06:00 AM  63983271000       639832   

   Incident Number   CAD Number  ...   Longitude  \
0        180006578  180031940.0  ... -122.428949   
1        186000748          NaN  ... -122.435866   
2        186034026          NaN  ... -122.472221   
3        186009552          NaN  ... -122.427631   
4        180151361  180570930.0  ... -122.411376   

                                  point SF Find Neighborhoods  \
0  (37.767504458676, -122.428948567562)                  28.0   
1  (37.798951849128, -122.435866321308)                  15.0   
2   (37.73946562952, -122.472220857192)                  46.0   
3  (37.774857948189, -122.427631000411)                  22.0   
4  (37.801125131493, -122.411376271773)                 107.0   

   Current Police Districts Current Supervisor Districts  \
0                       3.0                          5.0   
1                       4.0                          6.0   
2                      10.0                          8.0   
3                       4.0                         11.0   
4                       6.0                          3.0   

  Analysis Neighborhoods HSOC Zones as of 2018-06-05 OWED Public Spaces  \
0                    5.0                         5.0                NaN   
1                   13.0                         NaN                NaN   
2                   41.0                         NaN                NaN   
3                    9.0                         NaN                NaN   
4                    6.0                         NaN                NaN   

  Central Market/Tenderloin Boundary Polygon - Updated  \
0                                                NaN     
1                                                NaN     
2                                                NaN     
3                                                NaN     
4                                                NaN     

   Parks Alliance CPSI (27+TL sites)  
0                                NaN  
1                                NaN  
2                                NaN  
3                                NaN  
4                                NaN  

[5 rows x 34 columns]

Now that **Table 1** (**new_table_1**) and **Table 2** (**new_table_2**) are sorted by date, I am going to merge the two dataframes together. The rows will begin with reports starting in 2003 and will end with reports in 2019. All columns from **Table 2** will be kept for now.

### *TABLE 3: Merged data from Table 1 and Table 2* 

In [12]:
# Merged new_table_2 on new_table_1 to get a full table of all incident reports sorted by their dates
merged = new_table_1.merge(new_table_2, how='outer')
merged.head()

Incident Number       Incident Category  \
0        100087859  SEX OFFENSES, FORCIBLE   
1         91139945  SEX OFFENSES, FORCIBLE   
2         81360718            EMBEZZLEMENT   
3         81313927          SUSPICIOUS OCC   
4         81183231  FORGERY/COUNTERFEITING   

                    Incident Description Incident Day of Week Incident Date  \
0               ANNOY OR MOLEST CHILDREN            Wednesday    01/01/2003   
1                         SEXUAL BATTERY            Wednesday    01/01/2003   
2  EMBEZZLEMENT, GRAND THEFT BY EMPLOYEE            Wednesday    01/01/2003   
3                  SUSPICIOUS OCCURRENCE            Wednesday    01/01/2003   
4    FORGERY, DRIVERS LICENSE OR ID-CARD            Wednesday    01/01/2003   

  Incident Time Police District                              Resolution  \
0         00:01        SOUTHERN        COMPLAINANT REFUSES TO PROSECUTE   
1         00:01         BAYVIEW  DISTRICT ATTORNEY REFUSES TO PROSECUTE   
2         00:01        RICHMOND                                    NONE   
3         00:01         MISSION                                    NONE   
4         00:01        NORTHERN                                    NONE   

                      Address   Longitude  ...  Incident Code  \
0      800 Block of BRYANT ST -122.403405  ...            NaN   
1  200 Block of BRIDGEVIEW DR -122.397590  ...            NaN   
2    2900 Block of PACIFIC AV -122.443835  ...            NaN   
3    500 Block of GUERRERO ST -122.424037  ...            NaN   
4   3000 Block of FILLMORE ST -122.435517  ...            NaN   

  Incident Subcategory  Intersection  CNN  Analysis Neighborhood  \
0                  NaN           NaN  NaN                    NaN   
1                  NaN           NaN  NaN                    NaN   
2                  NaN           NaN  NaN                    NaN   
3                  NaN           NaN  NaN                    NaN   
4                  NaN           NaN  NaN                    NaN   

   Supervisor District  HSOC Zones as of 2018-06-05 OWED Public Spaces  \
0                  NaN                          NaN                NaN   
1                  NaN                          NaN                NaN   
2                  NaN                          NaN                NaN   
3                  NaN                          NaN                NaN   
4                  NaN                          NaN                NaN   

   Central Market/Tenderloin Boundary Polygon - Updated  \
0                                                NaN      
1                                                NaN      
2                                                NaN      
3                                                NaN      
4                                                NaN      

  Parks Alliance CPSI (27+TL sites)  
0                               NaN  
1                               NaN  
2                               NaN  
3                               NaN  
4                               NaN  

[5 rows x 35 columns]

## Defining the Data

Now that all the police reports are consolidated into one dataframe, it is important to define the data and to make it clear what each part of the dataframe represents. Firstly, the rows in the merged dataframe are individual reports for crimes that were officially documented in the San Franscisco Police Department, and the columns contain descriptive features of each report. The updated version after 2018 has some additional columns of information relating to each report, so the data from 2003 to 2017 will have some missing row values (represented as *NaN*).

In [13]:
merged.shape

(2415298, 35)

The merged dataframe has a total of 2,415,298 rows and 35 columns. Because there are over 2 million entries, it is important to take into account columns that may have a lot of missing values. Any additional columns from **Table 2** that aren't found in **Table 1** and vice versa will have missing data. However, **Table 1** has 2,215,024 entries, which takes up the majority of the merged dataframe (~92%). I plan on keeping all columns from **Table 1** that were merged with **Table 2** since there will be a sufficient amount of data to analyze. I want to examine the other columns to see which ones could be valuable before discarding any missing data because some of the information may be useful to look at more recent crime trends.

In [14]:
merged.columns

Index(['Incident Number', 'Incident Category', 'Incident Description',
       'Incident Day of Week', 'Incident Date', 'Incident Time',
       'Police District', 'Resolution', 'Address', 'Longitude', 'Latitude',
       'point', 'Row ID', 'SF Find Neighborhoods', 'Current Police Districts',
       'Current Supervisor Districts', 'Analysis Neighborhoods',
       'Incident Datetime', 'Incident Year', 'Report Datetime', 'Incident ID',
       'CAD Number', 'Report Type Code', 'Report Type Description',
       'Filed Online', 'Incident Code', 'Incident Subcategory', 'Intersection',
       'CNN', 'Analysis Neighborhood', 'Supervisor District',
       'HSOC Zones as of 2018-06-05', 'OWED Public Spaces',
       'Central Market/Tenderloin Boundary Polygon - Updated',
       'Parks Alliance CPSI (27+TL sites)'],
      dtype='object')

In [19]:
# Display a count of the number of missing values for each column in the merged table
merged.isnull().sum()

Incident Number                                               0
Incident Category                                            19
Incident Description                                          0
Incident Day of Week                                          0
Incident Date                                                 0
Incident Time                                                 0
Police District                                               1
Resolution                                                    0
Address                                                  246942
Longitude                                                 14315
Latitude                                                  14315
point                                                     14315
Row ID                                                        0
SF Find Neighborhoods                                     25205
Current Police Districts                                  15780
Current Supervisor Districts            

From the counts of the missing values in each column of the merged dataframe, it is clear that the columns with the most missing values are the ones that are unique to **Table 2**. **Table 2** is much smaller in size compared to **Table 1**, so it makes sense that majority of values in these columns would be missing. However, before performing any data imputation for these crime report features, I think it is important to examine the data that exists first. This is because the data that exists is mainly 2018-present day. I can observe this specific data to highlight more recent trends in San Francisco crime and highlight the potential differences from past crime trends.